Simplify child view logic.

jmcarp opened the pull request:
* Fetch html partials from `MainView`
* Don't pass callbacks from `MainView` to `ChildView` to `MetaModel`
* Consolidate redundant logic between `MainView` methods

Note: tests are expected to fail for now. I'll update and expand them if you all think the approach makes sense. cc @xtine @cmc333333 @theresaanna @ascott1 

[Resolves https://github.com/eregs/regulations-site/issues/289]

xtine commented:
I like the direction of the refactor, especially the simplification of some functions and setting options in a function. I'm assuming it's a WIP? Because testing switching between read/write modes breaks after a few switches (write mode doesn't activate). Didn't track down the event control in case you know what potentially could be causing it already.

jmcarp commented:
Yes, this is definitely a WIP, but also far enough along for review if folks have time. I also just pushed some updates that should fix some issues around switching preamble sections.

cmc333333 commented on regulations/static/regulations/js/source/models/meta-model.js:
```diff
@@ -51,73 +51,29 @@ var MetaModel = Backbone.Model.extend({
-        $deferred.resolve(this.content[id]);
-
-        return $deferred.promise();
+      return $.when(this.content[id]);
```
Neat

cmc333333 replied:
In the final pass, would you mind documenting this function? Something like: "The contents are already cached, so return them as a fulfilled promise"

Kinda related, would it make sense to either rename the function or just inline it? `retrieve()` doesn't convey enough information and could easily be confused with `request()`

cmc333333 commented on regulations/static/regulations/js/source/views/main/child-view.js:
```diff
@@ -14,56 +14,26 @@ Backbone.$ = $;
-            DrawerEvents.trigger('section:open', this.id);
+          if (this.title) {
+            this.title = this.assembleTitle();
+          }
```
This looks like the opposite of the above logic:
```javascript
                if (typeof this.title === 'undefined') {
                    this.title = this.assembleTitle();
                }
```

jmcarp replied:
Yes, will fix.

cmc333333 commented on regulations/static/regulations/js/source/views/main/child-view.js:
```diff
@@ -14,56 +14,26 @@ Backbone.$ = $;
-    setElement: function() {
-        if (this.id) {
-            Backbone.View.prototype.setElement.call(this, '#' + this.id);
-        }
```
I've followed along so far, but I'm not sure why this (and the associated `listenTo`) were deleted? What did they do and why is that no longer needed?

jmcarp replied:
As of `master`, the child view doesn't have access to the element it's managing when it's initialized. The child view's model calls a callback on partial load that indirectly calls `MainView.render`, which emits a `section:rendered` event, which lets the child know that its element now exists, and that it can look it up.

Under this proposal, we don't instantiate the child view until the partial has been loaded and added to the dom, so we don't need this indirection. We just have `MainView` pass the managed element to the constructor of the child view.

cmc333333 replied:
Is it safe to say that we were abusing `setElement` before and that we no longer need the special logic it contained?

cmc333333 commented on regulations/static/regulations/js/source/views/main/main-view.js:
```diff
@@ -32,74 +32,55 @@ var MainView = Backbone.View.extend({
+      if (this.contentType === 'search-results') {
+        options.params = URI.parseQuery(window.location.search);
+        options.query = options.params.q;
+      }
```
Should this be moved into `setChildOptions`, too?

jmcarp replied:
I don't think so--we call `setChildOptions` before instantiating a new child view, meaning we haven't pushed the new query parameters to the window history yet.

cmc333333 commented on regulations/static/regulations/js/source/views/main/reg-view.js:
```diff
@@ -95,21 +89,13 @@ var RegView = ChildView.extend({
+        return 'Appendix ' + this.id.substr(this.id.length - 1) + ' to Part ' + this.docId + ' | eRegulations';
+      } else {
+        return this.cfrTitle + ' CFR ' + Helpers.idToRef(this.id) + ' | eRegulations';
+      }
```
Excellent

cmc333333 commented on regulations/static/regulations/js/source/views/main/main-view.js:
```diff
@@ -124,61 +105,87 @@ var MainView = Backbone.View.extend({
+      }
+
+      this.childOptions.el = this.$el.children().get(0);
+      this.childView = new this.viewmap[this.contentType](this.childOptions);
```
Curious -- `renderView` and `createView` are both instantiating this `childView`. Why are they separate functions?

Somewhat related, in `createView` any existing `childView` is destroyed; do we need to do the same here?

cmc333333 replied:
Nevermind! Confused by views and models.

cmc333333 commented on regulations/static/regulations/js/source/views/main/main-view.js:
```diff
@@ -32,74 +32,55 @@ var MainView = Backbone.View.extend({
+      }
+
+      var options = {
+        el: '#' + this.sectionId,
```
Curious: `el` has been `#content-wrapper` for several of the child views. Why is it safe to move away from that?

jmcarp replied:
Interesting--because of the override on `setElement` in `ChildView` on `master`, setting `el` like this should have no effect. So it's not clear to me whether the intention of the current code is to set `el` to `#content-wrapper`, or to the id of the nested section--which typically aren't the same thing. Anyway, if we're willing to require that all child partials include a `#content-wrapper`, we should be able to stick with a fixed selector on `ChildView`.

cmc333333 commented on regulations/static/regulations/js/source/views/main/reg-view.js:
```diff
@@ -72,17 +66,17 @@ var RegView = ChildView.extend({
         }
 
-        if (id.length > 0) {
+        if (id) {
```
There's a risk here if the section id is zero, yes?

cmc333333 commented on regulations/static/regulations/js/source/views/main/preamble-view.js:
```diff
@@ -28,15 +26,12 @@ var PreambleView = ChildView.extend({
     var type = parsed.type;
 
-    this.id = [].concat(parsed.docId, parsed.type, parsed.section).join('-');
+    this.id = this.options.id;
```
Why can this logic be removed? Was it redundant?

cmc333333 commented:
Great stuff; this definitely removes a few steps of indirection

jmcarp commented:
Updated: fixed tests, addressed comments, and handled a few regressions I noticed during manual testing. I think this is ready for more review and QA if folks have time.

cmc333333 commented:
Hm. Ran into some issues locally:

1. Run the parser to load the data for `eregs pipeline 27 447`
2. Hard load `447-34/2015-12992`. 
3. Click the first definition, "articles"
4. Spinner keeps spinning
<img width="979" alt="screen shot 2016-04-28 at 1 04 25 pm" src="https://cloud.githubusercontent.com/assets/326918/14894468/c46a433e-0d41-11e6-9f03-9213e5851faa.png">

---

1. Do actions 1 and 2 above
2. Click 447.1 in the ToC
3. Click 447.2 in the ToC
<img width="1238" alt="screen shot 2016-04-28 at 1 03 59 pm" src="https://cloud.githubusercontent.com/assets/326918/14894448/b32a0258-0d41-11e6-8926-e2483a75d0bd.png">

---
Similar error:
1. Do actions 1 and 2 above
2. Enter a search term
3. Click result link
<img width="1245" alt="screen shot 2016-04-28 at 1 05 21 pm" src="https://cloud.githubusercontent.com/assets/326918/14894524/ffaa248c-0d41-11e6-8cad-93b5312bd79e.png">

jmcarp commented:
@cmc333333 thanks for catching. There were a few lingering callbacks getting passed around that should've been replaced with promises, which is done now.

cmc333333 commented:
Some errors while navigating diffs via the ToC
<img width="1146" alt="screen shot 2016-04-29 at 1 53 51 pm" src="https://cloud.githubusercontent.com/assets/326918/14924720/de63ac64-0e11-11e6-9dc5-1eee1c7b87fa.png">

Other things seem to be working well

cmc333333 commented on regulations/static/regulations/js/source/models/diff-model.js:
```diff
@@ -1,12 +1,22 @@
 
 var diffModel = new Backbone.DiffModel({
-    supplementalPath: 'diff'
+  getAJAXUrl: function(id, options) {
```
NBD, but a little odd that there's an `id` parameter that's never used.

cmc333333 commented:
Seems to be working well! Very glad to see this get in. I tested navigating the CFR, loading definitions, following links in the ToC and definition, following footer links, following links in the Diff ToC, etc.
